domain/src/main/java/org/oppia/domain/topic/TopicController.kt

```diff
@@ -186,6 +220,170 @@ class TopicController @Inject constructor(
     )
   }
 
+  fun getCompletedStoryList(profileId: ProfileId): LiveData<AsyncResult<CompletedStoryList>> {
+    return dataProviders.convertToLiveData(retrieveCompletedStoryListDataProvider(profileId))
+  }
+
+  fun getOngoingTopicList(profileId: ProfileId): LiveData<AsyncResult<OngoingTopicList>> {
+    return dataProviders.convertToLiveData(retrieveOngoingTopicListDataProvider(profileId))
+  }
+
+  private fun retrieveCompletedStoryListDataProvider(profileId: ProfileId): DataProvider<CompletedStoryList> {
+    return dataProviders.transformAsync(
+      TRANSFORMED_GET_COMPLETED_STORIES_PROVIDER_ID,
+      storyProgressController.retrieveCacheStore(profileId)
+    ) {
+      val completedStoryListBuilder = CompletedStoryList.newBuilder()
+      it.topicProgressMap.keys.forEach { topicId ->
```

Comment:
This could probably be done functionally, as well:

```kotlin
val completedStorySummaries = it.topicProgressMap.entries.flatMap { topicProgressEntry ->
  topicProgressEntry.value.storyProgressMap.mapKeys { storyProgressEntry ->
    retrieveStory(storyProgressEntry.key)
  }.filter(this::isStoryFullyCompleted).keys
}

...

private fun isStoryFullyCompleted(storySummaryAndProgress: Map.Entry<StorySummary, StoryProgress>): Boolean {
  val storySummary = storySummaryAndProgress.key
  val storyProgress = storySummaryAndProgress.value
  val chapterProgressMap = storyProgress.chapterProgressMap
  val lastChapterSummary = storySummary.chapterList.last()
  val lastExplorationId = lastChapterSummary.explorationId
  return lastExplorationId in chapterProgressMap && chapterProgressMap[lastExplorationId] == ChapterPlayState.COMPLETED
}
```

This approach avoids needing share mutable state across lambda boundaries (which might have future benefits, e.g. parallelizing the functional computations).

Reply:
Ditto below.

Reply:
Changed implementation which has made this a bit more simpler.